File: src/dictionary-loader.ts
import { AffixRule, AffixRules } from "./affix-rules";
import { Trie } from "./trie";

export type DictionarySource =
  | string
  | URL
  | (() => Promise<string>)
  | (() => string);

export class DictionaryLoader {
  /**
   * Load dictionary from various sources: file path, URL, or content provider function
   */
  static async loadDictionary(
    source: DictionarySource,
    trie: Trie,
  ): Promise<number> {
    const content = await this.getContent(source);
    return this.parseDictionaryContent(content, trie);
  }

  /**
   * Load affix rules from various sources: file path, URL, or content provider function
   */
  static async loadAffixRules(
    source: DictionarySource,
    affixRules: AffixRules,
  ): Promise<void> {
    const content = await this.getContent(source);
    this.parseAffixContent(content, affixRules);
  }

  /**
   * Synchronous version for backward compatibility with file paths
   */
  static loadDictionarySync(filePath: string, trie: Trie): number {
    if (typeof window !== "undefined") {
      throw new Error(
        "Synchronous file loading is not supported in browser environments. Use loadDictionary() instead.",
      );
    }

    // Use indirect require to avoid ESLint no-require-imports error
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const nodeRequire = (global as any).require || require;
    const { readFileSync } = nodeRequire("fs");
    const content = readFileSync(filePath, "utf8");
    return this.parseDictionaryContent(content, trie);
  }

  /**
   * Synchronous version for backward compatibility with file paths
   */
  static loadAffixRulesSync(filePath: string, affixRules: AffixRules): void {
    if (typeof window !== "undefined") {
      throw new Error(
        "Synchronous file loading is not supported in browser environments. Use loadAffixRules() instead.",
      );
    }

    // Use indirect require to avoid ESLint no-require-imports error
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const nodeRequire = (global as any).require || require;
    const { readFileSync } = nodeRequire("fs");
    const content = readFileSync(filePath, "utf8");
    this.parseAffixContent(content, affixRules);
  }

  private static async getContent(source: DictionarySource): Promise<string> {
    if (typeof source === "string") {
      // Check if it's a URL
      if (source.startsWith("http://") || source.startsWith("https://")) {
        const response = await fetch(source);
        if (!response.ok) {
          throw new Error(
            `Failed to fetch dictionary from ${source}: ${response.statusText}`,
          );
        }
        return response.text();
      }

      // Treat as file path - use dynamic import to avoid bundling fs in browser builds
      if (typeof window === "undefined") {
        const { readFileSync } = await import("fs");
        return readFileSync(source, "utf8");
      } else {
        throw new Error(
          "File path loading is not supported in browser environments. Use URL or content provider function instead.",
        );
      }
    }

    if (source instanceof URL) {
      const response = await fetch(source.toString());
      if (!response.ok) {
        throw new Error(
          `Failed to fetch dictionary from ${source}: ${response.statusText}`,
        );
      }
      return response.text();
    }

    if (typeof source === "function") {
      const result = source();
      return result instanceof Promise ? await result : result;
    }

    throw new Error(
      "Invalid dictionary source. Must be a file path, URL, or content provider function.",
    );
  }

  private static parseDictionaryContent(content: string, trie: Trie): number {
    const lines = content.split("\n");
    const wordCount = parseInt(lines[0], 10);

    for (let i = 1; i <= wordCount; i++) {
      if (lines[i]) {
        const [word] = lines[i].split("/");
        trie.insert(word.toLowerCase());
      }
    }

    return wordCount;
  }

  private static parseAffixContent(
    content: string,
    affixRules: AffixRules,
  ): void {
    const lines = content.split("\n");

    let currentRule: AffixRule | null = null;
    let expectedEntries = 0;

    for (const line of lines) {
      if (!line.trim() || line.startsWith("#")) continue;

      const parts = line.split(/\s+/);
      if (parts.length < 4) continue;

      if ((parts[0] === "PFX" || parts[0] === "SFX") && parts.length === 4) {
        const [type, flag, crossProduct, count] = parts;
        expectedEntries = parseInt(count, 10);
        currentRule = {
          type: type as "PFX" | "SFX",
          flag,
          crossProduct: crossProduct === "Y",
          entries: [],
        };
      } else if (
        currentRule &&
        parts[0] === currentRule.type &&
        parts[1] === currentRule.flag
      ) {
        const [, , strip, add, condition] = parts;
        currentRule.entries.push({
          strip: strip === "0" ? "" : strip,
          add,
          condition: condition === "." ? "" : condition,
        });

        if (currentRule.entries.length === expectedEntries) {
          affixRules.addRule(currentRule);
          currentRule = null;
          expectedEntries = 0;
        }
      }
    }
  }
}

File: rust/src/dictionary_loader.rs
use std::fs;
use std::io::{self, BufRead};
use std::path::Path;

use crate::affix_rules::{AffixEntry, AffixRule, AffixRules};
use crate::trie::Trie;

pub struct DictionaryLoader;

impl DictionaryLoader {
    pub fn load_dictionary<P: AsRef<Path>>(dic_path: P, trie: &mut Trie) -> io::Result<usize> {
        let file = fs::File::open(dic_path)?;
        let reader = io::BufReader::new(file);
        let mut lines = reader.lines();

        let word_count: usize = lines
            .next()
            .ok_or(io::Error::new(io::ErrorKind::InvalidData, "Missing word count"))??
            .parse()
            .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "Invalid word count"))?;

        for line in lines.take(word_count) {
            if let Ok(word_line) = line {
                let word = word_line.split('/').next().unwrap_or("");
                trie.insert(word);
            }
        }

        Ok(word_count)
    }

    pub fn load_affix_rules<P: AsRef<Path>>(aff_path: P, affix_rules: &mut AffixRules) -> io::Result<()> {
        let file = fs::File::open(aff_path)?;
        let reader = io::BufReader::new(file);

        let mut current_rule: Option<AffixRule> = None;
        let mut expected_entries = 0;

        for line in reader.lines() {
            let line = line?;
            if line.trim().is_empty() || line.starts_with('#') {
                continue;
            }

            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() < 4 {
                continue;
            }

            if (parts[0] == "PFX" || parts[0] == "SFX") && parts.len() == 4 {
                let (rule_type, flag, cross_product, count) = (parts[0], parts[1], parts[2], parts[3]);
                expected_entries = count.parse().unwrap_or(0);
                current_rule = Some(AffixRule {
                    rule_type: rule_type.to_string(),
                    flag: flag.to_string(),
                    cross_product: cross_product == "Y",
                    entries: Vec::new(),
                });
            } else if let Some(ref mut rule) = current_rule {
                if parts[0] == rule.rule_type && parts[1] == rule.flag {
                    let strip = if parts[2] == "0" { "" } else { parts[2] };
                    let add = parts[3];
                    let condition = parts.get(4).unwrap_or(&"").to_string();
                    rule.entries.push(AffixEntry {
                        strip: strip.to_string(),
                        add: add.to_string(),
                        condition,
                    });
                    expected_entries -= 1;

                    if expected_entries == 0 {
                        affix_rules.add_rule(rule.clone());
                        current_rule = None;
                    }
                }
            }
        }

        Ok(())
    }
}
